include/fmt/format.h

```diff
@@ -3283,6 +3261,21 @@ FMT_CONSTEXPR20 void format_hexfloat(Float value, int precision,
 template <typename Float>
 FMT_CONSTEXPR20 auto format_float(Float value, int precision, float_specs specs,
                                   buffer<char>& buf) -> int {
+  
+  // For checking rounding thresholds.
+  // The kth entry is chosen to be the smallest integer such that the
+  // upper 32-bits of 10^(k+1) times it is strictly bigger than 5 * 10^k.
+  static constexpr uint32_t fractional_part_rounding_thresholds[8] = {
```

Comment:
You need to drop static here:

```
/Users/runner/work/fmt/fmt/include/fmt/format.h:3268:29: error: definition of a static variable in a constexpr function is a C++2b extension [-Werror,-Wc++2b-extensions]
  static constexpr uint32_t fractional_part_rounding_thresholds[8] = {
                            ^
```

Also please move to a separate function to avoid parametrization on `Float`, something like

```c++
FMT_CONSTEXPR uint32_t fractional_part_rounding_threshold(int index) {
  constexpr uint32_t thresholds[] = {...};
  return thresholds[index];
}
```

Reply:
@vitaut 
Will a non-static constexpr array be optimized by the compiler?

Reply:
¯\_(ツ)_/¯

Reply:
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2647r1.html#workarounds
So, yes.
¯_(ツ)_/¯

Reply:
Done.

Reply:
@mwinterb So the linked article says that it doesn't get optimized, no?

Reply:
It doesn't get optimized: https://godbolt.org/z/drn31x9zK

Reply:
But clang will:
https://godbolt.org/z/YdYMGqs46
Sorry. My "So, yes" was agreeing with the shrug. "maybe, maybe not".
Should have been more clear.

Reply:
> But clang will: https://godbolt.org/z/YdYMGqs46 Sorry. My "So, yes" was agreeing with the shrug. "maybe, maybe not". Should have been more clear.

Really sorry about that confusion.

Reply:
@mwinterb Ah, I see, my bad 😋

Reply:
@vitaut 

> It doesn't get optimized: https://godbolt.org/z/drn31x9zK

Maybe revert this PR?

Reply:
I don't think we should revert this PR but ideas/PRs to optimize it would be welcome.

Reply:
@vitaut 
Solution for GCC, Clang:  https://godbolt.org/z/ezj9GhnYW
But not for MSVC (current code: https://godbolt.org/z/5Pxn3zMfT)

Reply:
> I don't think we should revert this PR but ideas/PRs to optimize it would be welcome.

Turn it into a switch:  https://godbolt.org/z/YPj6G3zPd

Reply:
I have another idea: https://github.com/fmtlib/fmt/commit/06b20387aebdd47444b9eb3d3b3bc254251fca3b.

Reply:
> @vitaut Solution for GCC, Clang: https://godbolt.org/z/ezj9GhnYW But not for MSVC (current code: https://godbolt.org/z/5Pxn3zMfT)

That looks good.  I had something similar but didn't think of the `unreachable` to quiet the warnings and keep the optimization.